src/java/org/apache/cassandra/metrics/CassandraMetricsRegistry.java

```diff
@@ -444,30 +410,71 @@ void remove(ThreadPoolMetrics metrics)
 
     public <T extends Metric> T register(MetricName name, T metric, MetricName... aliases)
     {
-        setAliases(ArrayUtils.addAll(new MetricName[]{name}, aliases));
         T metricLoc = register(name, metric);
         Stream.of(aliases).forEach(n -> register(n, metricLoc));
         return metricLoc;
     }
 
+    /**
+     * Removes all metrics that match the given predicate.
+     *
+     * @param resolver a function that resolves a short metric name from a full metric name,
+     * or @{code null} if the metric should not be removed
+     * @param factory a function that creates a metric name from a short metric name.
+     * @param onRemoved a callback that is called for each removed metric.
+     */
+    public void removeIfMatch(MetricNameResolver resolver,
+                              Function<String, MetricName> factory,
+                              Consumer<MetricName> onRemoved)
+    {
+        removeMatching((full, metric) -> {
+            String shortName = resolver.resolve(full);
+            if (shortName == null)
+                return false;
+
+            MetricName metricName = factory.apply(shortName);
+            boolean remove = metricName.getMetricName().equals(full);
+
+            if (remove)
+            {
+                unregisterMBean(metricName.getMBeanName(), MBeanWrapper.instance);
+                onRemoved.accept(metricName);
+            }
+            return remove;
+        });
+    }
+
+    /**
+     * Default implementation of the {@link MetricNameResolver} that resolves a short metric name from a full metric name,
+     * assuming that the full metric name doesn't contain dots. Returns {@code null} if the full metric name doesn't match
+     * the provided group and type.
+     *
+     * @param fullName full metric name
+     * @param group metric group
+     * @param type metric type
+     * @return short metric name or {@code null} if the full metric name doesn't match the group and type.
+     */
+    public static @Nullable String resolveShortMetricName(String fullName, String group, String type)
+    {
+        String prefix = name(group, type);
+        if (fullName.startsWith(prefix))
+            return fullName.substring(prefix.length() + 1, fullName.indexOf('.', prefix.length() + 1));
+
+        return null;
+    }
+
     public void remove(MetricName name)
     {
         // Aliases are removed in onMetricRemoved by metrics listener.
```

Comment:
nit: This comment is no longer true?

Reply:
The way this works does make more sense now, though. The table metrics were the only place we cared about the aliases, and they handle removal specifically now.

Reply:
Comment removed.